Make complex its own backend

li-roy opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #19295 Generate only one Type class per backend
* **#19275 Make complex its own backend**
* #19270 Add ScalarType argument to Type::options()
* #19230 Generate cases for all ScalarTypes in Type functions

Differential Revision: [D14948582](https://our.internmc.facebook.com/intern/diff/D14948582)

Commit:
Make complex its own backend

Commit:
Update on "Make complex its own backend"

Make complex its own backend

gh-metadata: pytorch pytorch 19275 gh/li-roy/10/head

Commit:
Update on "Make complex its own backend"

Make complex its own backend

gh-metadata: pytorch pytorch 19275 gh/li-roy/10/head

Commit:
Update on "Make complex its own backend"

Make complex its own backend

gh-metadata: pytorch pytorch 19275 gh/li-roy/10/head

Commit:
Update on "Make complex its own backend"

Make complex its own backend

gh-metadata: pytorch pytorch 19275 gh/li-roy/10/head

Commit:
Update on "Make complex its own backend"

Make complex its own backend

gh-metadata: pytorch pytorch 19275 gh/li-roy/10/head

Commit:
Update on "Make complex its own backend"

Make complex its own backend

gh-metadata: pytorch pytorch 19275 gh/li-roy/10/head

Commit:
Update on "Make complex its own backend"

Make complex its own backend

gh-metadata: pytorch pytorch 19275 gh/li-roy/10/head

Commit:
Update on "Make complex its own backend"

Make complex its own backend

gh-metadata: pytorch pytorch 19275 gh/li-roy/10/head

Commit:
Update on "Make complex its own backend"

Make complex its own backend

gh-metadata: pytorch pytorch 19275 gh/li-roy/10/head

Commit:
Update on "Make complex its own backend"

Make complex its own backend

gh-metadata: pytorch pytorch 19275 gh/li-roy/10/head

Commit:
Update on "Make complex its own backend"

Make complex its own backend

gh-metadata: pytorch pytorch 19275 gh/li-roy/10/head

dzhulgakov commented:
@gchanan @ezyang - can you double-check please? complex is magical

ezyang commented:
One general question I have, is that we have this standing recommendation from https://fb.quip.com/srjaAPbCP6Hg:

> * Eliminate Backend in favor of TensorTypeId
>    * Delete the backend() option from TensorOptions
>    * Need some replacement for TensorOptions-based dispatch
>    * There may be a lot of code that is typing tensor.type().backend() (Claim: They just meant device type.)

Did we decide not do this after? Would it make sense to do that first before adding some more backends?

ezyang commented on torch/csrc/utils/tensor_types.cpp:
```diff
@@ -20,6 +20,9 @@ static const char* backend_to_string(const at::Type& type) {
     case at::Backend::SparseCUDA: return "torch.cuda.sparse";
+    // We split complex into its own backend, but keeping it the same here for now
+    case at::Backend::ComplexCPU: return "torch";
+    case at::Backend::ComplexCUDA: return "torch.cuda";
```
There's only one use site of this function, in `type_to_string`. I see you have adjusted it in #19295 but the adjustment looks a bit questionable. Wouldn't it make more sense to stop having a "type to string" operation, and instead, just pass the trinity (device type, layout, dtype) and render that way?

ezyang replied:
Basically, after your change here, this function doesn't make "semantic" sense anymore. So I'd much prefer if we got rid of this function entirely. Can be a PR on top, if necessary.

ezyang commented on test/cpp_extensions/complex_registration_extension.cpp:
```diff
@@ -39,17 +39,35 @@ struct CPUComplexFloatType : public at::CPUTypeDefault {
+    AT_ASSERT(options.device().is_cpu());
+
+    for (auto x: size) {
+      AT_CHECK(x >= 0, "Trying to create tensor with negative dimension ", x, ": ", size);
```
Umm... x isn't the dimension index, it's the size of the dimension. You can probably just drop `x` entirely; the printing of `size` is good enough

ezyang commented on c10/core/TensorOptions.h:
```diff
@@ -357,8 +357,18 @@ struct C10_API TensorOptions {
+            // This is a hack to make TensorOptions based dispatch work for
+            // complex backends.
+            // TODO: Find a better solution and extend it to other backends that
+            // share the same Layout and DeviceType
```
I don't see what is wrong with this. `computeTensorTypeId` takes the tuple (device type, layout, dtype) and determines what the appropriate dispatch key is. It special cases strided, cpu, etc.; why can't it special case dtype? The only trouble is if you want *open* dtype/layout registration, but this situation wouldn't apply to complex anyway since it's already in our built-in enum.

ezyang commented:
Test failures look serious.

ezyang reviewed (approved):
Some follow up stuff to do but this diff looks basically ok.

Commit:
Update on "Make complex its own backend"

Make complex its own backend

gh-metadata: pytorch pytorch 19275 gh/li-roy/10/head

Commit:
Update on "Make complex its own backend"

Make complex its own backend

gh-metadata: pytorch pytorch 19275 gh/li-roy/10/head

Commit:
Update on "Make complex its own backend"

Make complex its own backend

gh-metadata: pytorch pytorch 19275 gh/li-roy/10/head

Commit:
Update on "Make complex its own backend"

Make complex its own backend

gh-metadata: pytorch pytorch 19275 gh/li-roy/10/head

Commit:
Update on "Make complex its own backend"

Make complex its own backend

gh-metadata: pytorch pytorch 19275 gh/li-roy/10/head

Commit:
Update on "Make complex its own backend"

Make complex its own backend

gh-metadata: pytorch pytorch 19275 gh/li-roy/10/head
